"multiple matching crates for `serde_derive`" in test `used_underscore_binding_macro`

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

For bug reports and false positives, please include the output of `cargo clippy -V` in the report.

Thank you for using Clippy!

Write your comment below this line: -->

Having applied https://github.com/rust-lang/rust/pull/57435 [manually](https://github.com/xftroxgpx/a3/blob/f45acfa61559a3ab381df0f4dfd3c0b06e710754/system/Z575/OSes/3archlinux/on_baremetal/filesystem_now/archlinux/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/57435.patch) (since it's not yet merged), and I don't know if this matters at all, I am hitting this issue locally (with clippy [HEAD](https://github.com/rust-lang/rust-clippy/commits/5b8496603c4db231719b67c1f88b409e7b278e5e) and rustc [HEAD](https://github.com/rust-lang/rust/commit/9d54812829e9d92dac35a4a0f358cdc5a2475371) compiled with [`debug-assertions = true`](https://github.com/xftroxgpx/a3/blob/f45acfa61559a3ab381df0f4dfd3c0b06e710754/system/Z575/OSes/3archlinux/on_baremetal/filesystem_now/archlinux/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/config.toml#L346) ), as follows:

```bash
/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy 
$ git pull && cargo clean && time cargo install --path . --force && cargo test --release
...
(same error here as below)
...
/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy 
$ time RUST_BACKTRACE=1 cargo test --release
!! LD_LIBRARY_PATH=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib:/home/xftroxgpx/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib:
!! Executing '/home/xftroxgpx/build/2nonpkgs/rust.stuff/cargo/cargo//target/release//cargo' in pwd='/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy' with args: 'test --release'
    Finished release [optimized] target(s) in 0.19s
     Running target/release/deps/compile_test-74bdeebcb1154d92

running 1 test

running 30 tests
test [run-pass] run-pass/ice-1588.rs ... ok
test [run-pass] run-pass/cc_seme.rs ... ok
test [run-pass] run-pass/associated-constant-ice.rs ... ok
test [run-pass] run-pass/enum-glob-import-crate.rs ... ok
test [run-pass] run-pass/ice-1969.rs ... ok
test [run-pass] run-pass/ice-1782.rs ... ok
test [run-pass] run-pass/ice-2594.rs ... ok
test [run-pass] run-pass/ice-2499.rs ... ok
test [run-pass] run-pass/ice-2760.rs ... ok
test [run-pass] run-pass/ice-2727.rs ... ok
test [run-pass] run-pass/ice-2865.rs ... ok
test [run-pass] run-pass/ice-2774.rs ... ok
test [run-pass] run-pass/ice-3462.rs ... ok
test [run-pass] run-pass/ice-700.rs ... ok
test [run-pass] run-pass/ice-3151.rs ... ok
test [run-pass] run-pass/ice_exacte_size.rs ... ok
test [run-pass] run-pass/if_same_then_else.rs ... ok
test [run-pass] run-pass/issue-2862.rs ... ok
test [run-pass] run-pass/issue-825.rs ... ok
test [run-pass] run-pass/issues_loop_mut_cond.rs ... ok
test [run-pass] run-pass/match_same_arms_const.rs ... ok
test [run-pass] run-pass/mut_mut_macro.rs ... ok
test [run-pass] run-pass/needless_lifetimes_impl_trait.rs ... ok
test [run-pass] run-pass/needless_borrow_fp.rs ... ok
test [run-pass] run-pass/returns.rs ... ok
test [run-pass] run-pass/regressions.rs ... ok
test [run-pass] run-pass/procedural_macro.rs ... ok
test [run-pass] run-pass/single-match-else.rs ... ok
test [run-pass] run-pass/used_underscore_binding_macro.rs ... FAILED
test [run-pass] run-pass/whitelist/conf_whitelisted.rs ... ok

failures:

---- [run-pass] run-pass/used_underscore_binding_macro.rs stdout ----

error: compilation failed!
status: exit code: 1
command: "target/release/clippy-driver" "tests/run-pass/used_underscore_binding_macro.rs" "-L" "/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/debug/test_build_base" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/debug/test_build_base/used_underscore_binding_macro.stage-id" "-L" "target/release" "-L" "target/release/deps" "-Dwarnings" "-Zui-testing" "-L" "/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/debug/test_build_base/used_underscore_binding_macro.stage-id.aux"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
{"message":"multiple matching crates for `serde_derive`","code":{"code":"E0464","explanation":null},"level":"error","spans":[{"file_name":"tests/run-pass/used_underscore_binding_macro.rs","byte_start":496,"byte_end":522,"line_start":13,"line_end":13,"column_start":1,"column_end":27,"is_primary":true,"text":[{"text":"extern crate serde_derive;","highlight_start":1,"highlight_end":27}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"candidates:\ncrate `serde_derive`: /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/release/deps/libserde_derive-e15228d7a19882ad.so\ncrate `serde_derive`: /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/release/deps/libserde_derive-3c0db970ced9d85e.so","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error[E0464]: multiple matching crates for `serde_derive`\n  --> tests/run-pass/used_underscore_binding_macro.rs:13:1\n   |\nLL | extern crate serde_derive;\n   | ^^^^^^^^^^^^^^^^^^^^^^^^^^\n   |\n   = note: candidates:\n           crate `serde_derive`: /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/release/deps/libserde_derive-e15228d7a19882ad.so\n           crate `serde_derive`: /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust-clippy/target/release/deps/libserde_derive-3c0db970ced9d85e.so\n\n"}
{"message":"can't find crate for `serde_derive`","code":{"code":"E0463","explanation":"\nA plugin/crate was declared but cannot be found. Erroneous code example:\n\n```compile_fail,E0463\n#![feature(plugin)]\n#![plugin(cookie_monster)] // error: can't find crate for `cookie_monster`\nextern crate cake_is_a_lie; // error: can't find crate for `cake_is_a_lie`\n```\n\nYou need to link your code to the relevant crate in order to be able to use it\n(through Cargo or the `-L` option of rustc example). Plugins are crates as\nwell, and you link to them the same way.\n"},"level":"error","spans":[{"file_name":"tests/run-pass/used_underscore_binding_macro.rs","byte_start":496,"byte_end":522,"line_start":13,"line_end":13,"column_start":1,"column_end":27,"is_primary":true,"text":[{"text":"extern crate serde_derive;","highlight_start":1,"highlight_end":27}],"label":"can't find crate","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error[E0463]: can't find crate for `serde_derive`\n  --> tests/run-pass/used_underscore_binding_macro.rs:13:1\n   |\nLL | extern crate serde_derive;\n   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate\n\n"}
{"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
{"message":"Some errors occurred: E0463, E0464.","code":null,"level":"","spans":[],"children":[],"rendered":"Some errors occurred: E0463, E0464.\n"}
{"message":"For more information about an error, try `rustc --explain E0463`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about an error, try `rustc --explain E0463`.\n"}

------------------------------------------

thread '[run-pass] run-pass/used_underscore_binding_macro.rs' panicked at 'explicit panic', /home/xftroxgpx/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.18/src/runtest.rs:2632:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: compiletest_rs::runtest::ProcRes::fatal
   7: compiletest_rs::runtest::TestCx::fatal_proc_rec
   8: compiletest_rs::runtest::TestCx::run_rpass_test
   9: compiletest_rs::runtest::run
  10: <F as alloc::boxed::FnBox<A>>::call_box
  11: <F as alloc::boxed::FnBox<A>>::call_box
  12: __rust_maybe_catch_panic
  13: std::panicking::try


failures:
    [run-pass] run-pass/used_underscore_binding_macro.rs

test result: FAILED. 29 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

test compile_test ... FAILED

failures:

---- compile_test stdout ----
thread 'compile_test' panicked at 'Some tests failed', /home/xftroxgpx/.cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.18/src/lib.rs:90:22
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: compiletest_rs::run_tests
   7: core::ops::function::FnOnce::call_once
   8: <F as alloc::boxed::FnBox<A>>::call_box
   9: __rust_maybe_catch_panic
  10: std::panicking::try


failures:
    compile_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--test compile-test'

real	0m8.983s
user	0m25.870s
sys	0m4.080s

```

Similar issue but for `serde` not `serde-derive` is: https://github.com/rust-lang/rust-clippy/issues/2896

--- a/.cargo/config
+++ b/.cargo/config
@@ -5,4 +5,5 @@ lintcheck = "run --target-dir lintcheck/target --package lintcheck --bin lintche
 collect-metadata = "test --test dogfood --features metadata-collector-lint -- run_metadata_collection_lint --ignored"
 
 [build]
-rustflags = ["-Zunstable-options"]
+# -Zbinary-dep-depinfo allows us to track which rlib files to use for compiling UI tests
+rustflags = ["-Zunstable-options", "-Zbinary-dep-depinfo"]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -32,11 +32,7 @@ tempfile = { version = "3.1.0", optional = true }
 cargo_metadata = "0.12"
 compiletest_rs = { version = "0.6.0", features = ["tmp"] }
 tester = "0.9"
-serde = { version = "1.0", features = ["derive"] }
-derive-new = "0.5"
 regex = "1.4"
-quote = "1"
-syn = { version = "1", features = ["full"] }
 # This is used by the `collect-metadata` alias.
 filetime = "0.2"
 
@@ -44,6 +40,15 @@ filetime = "0.2"
 # See the `src/tools/rustc-workspace-hack/README.md` file in `rust-lang/rust`
 # for more information.
 rustc-workspace-hack = "1.0.0"
+
+# UI test dependencies
+clippy_utils = { path = "clippy_utils" }
+derive-new = "0.5"
+if_chain = "1.0"
+itertools = "0.10.1"
+quote = "1"
+serde = { version = "1.0", features = ["derive"] }
+syn = { version = "1", features = ["full"] }
 
 [build-dependencies]
 rustc_tools_util = { version = "0.2.0", path = "rustc_tools_util" }

--- a/clippy_utils/Cargo.toml
+++ b/clippy_utils/Cargo.toml
@@ -6,10 +6,6 @@ publish = false
 
 [dependencies]
 if_chain = "1.0.0"
-itertools = "0.9"
-regex-syntax = "0.6"
-serde = { version = "1.0", features = ["derive"] }
-unicode-normalization = "0.1"
 rustc-semver="1.1.0"
 
 [features]

--- a/tests/compile-test.rs
+++ b/tests/compile-test.rs
@@ -1,5 +1,7 @@
 #![feature(test)] // compiletest_rs requires this attribute
 #![feature(once_cell)]
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
 
 use compiletest_rs as compiletest;
 use compiletest_rs::common::Mode as TestMode;
@@ -31,11 +33,17 @@ static TEST_DEPENDENCIES: &[&str] = &[
 
 // Test dependencies may need an `extern crate` here to ensure that they show up
 // in the depinfo file (otherwise cargo thinks they are unused)
+#[allow(unused_extern_crates)]
 extern crate clippy_utils;
+#[allow(unused_extern_crates)]
 extern crate derive_new;
+#[allow(unused_extern_crates)]
 extern crate if_chain;
+#[allow(unused_extern_crates)]
 extern crate itertools;
+#[allow(unused_extern_crates)]
 extern crate quote;
+#[allow(unused_extern_crates)]
 extern crate syn;
 
 fn host_lib() -> PathBuf {

--- a/tests/compile-test.rs
+++ b/tests/compile-test.rs
@@ -1,10 +1,10 @@
 #![feature(test)] // compiletest_rs requires this attribute
 #![feature(once_cell)]
-#![feature(try_blocks)]
 
 use compiletest_rs as compiletest;
 use compiletest_rs::common::Mode as TestMode;
 
+use std::collections::HashMap;
 use std::env::{self, remove_var, set_var, var_os};
 use std::ffi::{OsStr, OsString};
 use std::fs;
@@ -16,6 +16,28 @@ mod cargo;
 // whether to run internal tests or not
 const RUN_INTERNAL_TESTS: bool = cfg!(feature = "internal-lints");
 
+/// All crates used in UI tests are listed here
+static TEST_DEPENDENCIES: &[&str] = &[
+    "clippy_utils",
+    "derive_new",
+    "if_chain",
+    "itertools",
+    "quote",
+    "regex",
+    "serde",
+    "serde_derive",
+    "syn",
+];
+
+// Test dependencies may need an `extern crate` here to ensure that they show up
+// in the depinfo file (otherwise cargo thinks they are unused)
+extern crate clippy_utils;
+extern crate derive_new;
+extern crate if_chain;
+extern crate itertools;
+extern crate quote;
+extern crate syn;
+
 fn host_lib() -> PathBuf {
     option_env!("HOST_LIBS").map_or(cargo::CARGO_TARGET_DIR.join(env!("PROFILE")), PathBuf::from)
 }
@@ -24,72 +46,58 @@ fn clippy_driver_path() -> PathBuf {
     option_env!("CLIPPY_DRIVER_PATH").map_or(cargo::TARGET_LIB.join("clippy-driver"), PathBuf::from)
 }
 
-// When we'll want to use `extern crate ..` for a dependency that is used
-// both by the crate and the compiler itself, we can't simply pass -L flags
-// as we'll get a duplicate matching versions. Instead, disambiguate with
-// `--extern dep=path`.
-// See https://github.com/rust-lang/rust-clippy/issues/4015.
-//
-// FIXME: We cannot use `cargo build --message-format=json` to resolve to dependency files.
-//        Because it would force-rebuild if the options passed to `build` command is not the same
-//        as what we manually pass to `cargo` invocation
-fn third_party_crates() -> String {
-    use std::collections::HashMap;
-    static CRATES: &[&str] = &[
-        "clippy_lints",
-        "clippy_utils",
-        "if_chain",
-        "itertools",
-        "quote",
-        "regex",
-        "serde",
-        "serde_derive",
-        "syn",
-    ];
-    let dep_dir = cargo::TARGET_LIB.join("deps");
-    let mut crates: HashMap<&str, Vec<PathBuf>> = HashMap::with_capacity(CRATES.len());
-    let mut flags = String::new();
-    for entry in fs::read_dir(dep_dir).unwrap().flatten() {
-        let path = entry.path();
-        if let Some(name) = try {
-            let name = path.file_name()?.to_str()?;
-            let (name, _) = name.strip_suffix(".rlib")?.strip_prefix("lib")?.split_once('-')?;
-            CRATES.iter().copied().find(|&c| c == name)?
-        } {
-            flags += &format!(" --extern {}={}", name, path.display());
-            crates.entry(name).or_default().push(path.clone());
+/// Produces a string with an `--extern` flag for all UI test crate
+/// dependencies.
+///
+/// The dependency files are located by parsing the depinfo file for this test
+/// module. This assumes the `-Z binary-dep-depinfo` flag is enabled. All test
+/// dependencies must be added to Cargo.toml at the project root. Test
+/// dependencies that are not *directly* used by this test module require an
+/// `extern crate` declaration.
+fn extern_flags() -> String {
+    let current_exe_depinfo = {
+        let mut path = env::current_exe().unwrap();
+        path.set_extension("d");
+        std::fs::read_to_string(path).unwrap()
+    };
+    let mut crates: HashMap<&str, &str> = HashMap::with_capacity(TEST_DEPENDENCIES.len());
+    for line in current_exe_depinfo.lines() {
+        // each dependency is expected to have a Makefile rule like `/path/to/crate-hash.rlib:`
+        let parse_name_path = || {
+            if line.starts_with(char::is_whitespace) {
+                return None;
+            }
+            let path_str = line.strip_suffix(':')?;
+            let path = Path::new(path_str);
+            if !matches!(path.extension()?.to_str()?, "rlib" | "so" | "dylib" | "dll") {
+                return None;
+            }
+            let (name, _hash) = path.file_stem()?.to_str()?.rsplit_once('-')?;
+            // the "lib" prefix is not present for dll files
+            let name = name.strip_prefix("lib").unwrap_or(name);
+            Some((name, path_str))
+        };
+        if let Some((name, path)) = parse_name_path() {
+            if TEST_DEPENDENCIES.contains(&name) {
+                // A dependency may be listed twice if it is available in sysroot,
+                // and the sysroot dependencies are listed first. As of the writing,
+                // this only seems to apply to if_chain.
+                crates.insert(name, path);
+            }
         }
     }
-    crates.retain(|_, paths| paths.len() > 1);
-    if !crates.is_empty() {
-        let crate_names = crates.keys().map(|s| format!("`{}`", s)).collect::<Vec<_>>().join(", ");
-        // add backslashes for an easy copy-paste `rm` command
-        let paths = crates
-            .into_values()
-            .flatten()
-            .map(|p| strip_current_dir(&p).display().to_string())
-            .collect::<Vec<_>>()
-            .join(" \\\n");
-        // Check which action should be done in order to remove compiled deps.
-        // If pre-installed version of compiler is used, `cargo clean` will do.
-        // Otherwise (for bootstrapped compiler), the dependencies directory
-        // must be removed manually.
-        let suggested_action = if std::env::var_os("RUSTC_BOOTSTRAP").is_some() {
-            "removing the stageN-tools directory"
-        } else {
-            "running `cargo clean`"
-        };
-
-        panic!(
-            "\n----------------------------------------------------------------------\n\
-            ERROR: Found multiple rlibs for crates: {}\n\
-            Try {} or remove the following files:\n\n{}\n\n\
-            For details on this error see https://github.com/rust-lang/rust-clippy/issues/7343\n\
-            ----------------------------------------------------------------------\n",
-            crate_names, suggested_action, paths
-        );
+    let not_found: Vec<&str> = TEST_DEPENDENCIES
+        .iter()
+        .copied()
+        .filter(|n| !crates.contains_key(n))
+        .collect();
+    if !not_found.is_empty() {
+        panic!("dependencies not found in depinfo: {:?}", not_found);
     }
-    flags
+    crates
+        .into_iter()
+        .map(|(name, path)| format!("--extern {}={} ", name, path))
+        .collect()
 }
 
 fn default_config() -> compiletest::Config {
@@ -105,11 +113,14 @@ fn default_config() -> compiletest::Config {
         config.compile_lib_path = path;
     }
 
+    // Using `-L dependency={}` enforces that external dependencies are added with `--extern`.
+    // This is valuable because a) it allows us to monitor what external dependencies are used
+    // and b) it ensures that conflicting rlibs are resolved properly.
     config.target_rustcflags = Some(format!(
-        "--emit=metadata -L {0} -L {1} -Dwarnings -Zui-testing {2}",
+        "--emit=metadata -L dependency={} -L dependency={} -Dwarnings -Zui-testing {}",
         host_lib().join("deps").display(),
         cargo::TARGET_LIB.join("deps").display(),
-        third_party_crates(),
+        extern_flags(),
     ));
 
     config.build_base = host_lib().join("test_build_base");
@@ -315,13 +326,4 @@ impl Drop for VarGuard {
             Some(value) => set_var(self.key, value),
         }
     }
-}
-
-fn strip_current_dir(path: &Path) -> &Path {
-    if let Ok(curr) = env::current_dir() {
-        if let Ok(stripped) = path.strip_prefix(curr) {
-            return stripped;
-        }
-    }
-    path
 }

--- a/tests/dogfood.rs
+++ b/tests/dogfood.rs
@@ -6,6 +6,8 @@
 // Dogfood cannot run on Windows
 #![cfg(not(windows))]
 #![feature(once_cell)]
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
 
 use std::lazy::SyncLazy;
 use std::path::PathBuf;

--- a/tests/fmt.rs
+++ b/tests/fmt.rs
@@ -1,3 +1,6 @@
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
+
 use std::path::PathBuf;
 use std::process::Command;
 

--- a/tests/integration.rs
+++ b/tests/integration.rs
@@ -1,4 +1,6 @@
 #![cfg(feature = "integration")]
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
 
 use std::env;
 use std::ffi::OsStr;

--- a/tests/lint_message_convention.rs
+++ b/tests/lint_message_convention.rs
@@ -1,3 +1,6 @@
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
+
 use std::ffi::OsStr;
 use std::path::PathBuf;
 

--- a/tests/missing-test-files.rs
+++ b/tests/missing-test-files.rs
@@ -1,3 +1,5 @@
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
 #![allow(clippy::assertions_on_constants)]
 
 use std::fs::{self, DirEntry};

--- a/tests/versioncheck.rs
+++ b/tests/versioncheck.rs
@@ -1,4 +1,7 @@
+#![cfg_attr(feature = "deny-warnings", deny(warnings))]
+#![warn(rust_2018_idioms, unused_lifetimes)]
 #![allow(clippy::single_match_else)]
+
 use rustc_tools_util::VersionInfo;
 
 #[test]

